service/matching/version_rule_helpers.go

```diff
@@ -0,0 +1,244 @@
+package matching
+
+import (
+	"fmt"
+	"slices"
+	"time"
+
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/api/taskqueue/v1"
+	"go.temporal.io/api/workflowservice/v1"
+
+	persistencepb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common"
+	hlc "go.temporal.io/server/common/clock/hybrid_logical_clock"
+	"go.temporal.io/server/common/util"
+)
+
+// checkAssignmentConditions returns an error if the new set of assignment rules don't meet the following requirements:
+// - No more rules than dynamicconfig.VersionAssignmentRuleLimitPerQueue
+// - Once assignment rules have been added, there must always be at least one "unfiltered" assignment rule, which can accept any build id
```

Comment:
I think the desired behavior is to enforce unfiltered rule only if there was an unfiltered rule prior to the update. consider the case that users transitioning from unversioned to versioned and want's to do it gradually, during the transition there is no unfiltered rules.

Reply:
ok that's nicer I think. and we can support it. I'll make the change